### PR TITLE
Kusto trusted endpoints

### PR DESCRIPTION
**Pull request by ohadbitt**
Validate endpoint before executing query or streaming

### CODE REVIEW

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    lst = copy.deepcopy(rules)
+    for _, v in existing.rules.items():
+        lst = lst + v
+    return FastSuffixMatcher(lst)
+
+
+class KustoTrustedEndpoints:
+    def __init__(self):
+        self._matchers = dict()
```
Comment (AsafMah) [inline azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py:62]: A one-liner if you want, as is tradition:
```python
        self._matchers = {
            k: FastSuffixMatcher(
                [*(MatchRule(suffix, False) for suffix in v["AllowedKustoSuffixes"]), *(MatchRule(hostname, True) for hostname in v["AllowedKustoHostnames"])]
            )
            for (k, v) in _well_known_kusto_endpoints_data["AllowedEndpointsByLogin"].items()
        }
```
(Only if it's readable for you)

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        for k, v in _well_known_kusto_endpoints_data["AllowedEndpointsByLogin"].items():
+            rules = itertools.chain(
```
Comment (AsafMah) [inline azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py:64]: Iterators are a bit overkill for a small list of known size. If it's just to combine them, use splats:

```python
[*(MatchRule(suffix, False) for suffix in v["AllowedKustoSuffixes"]), *(MatchRule(hostname, True) for hostname in v["AllowedKustoHostnames"])]
```

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+class MatchRule:
+    def __init__(self, suffix, exact):
+        self.suffix = suffix.lower()
+        self.exact = exact
+
+
+class FastSuffixMatcher:
+    def __init__(self, rules: Iterable[MatchRule]):
+        it1, it2 = itertools.tee(rules)
```
Comment (AsafMah) [inline azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py:22]: If it was a list and not an iterator, you wouldn't need this trick (see below)

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py (hunk 1 above)
Comment (ohadbitt) [reply azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py:62]: it is beautiful

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py (hunk 2 above)
Comment (ohadbitt) [reply azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py:64]: thanks its so much better

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        self._suffix_length = min([len(rule.suffix) for rule in it1])
```
[the remark this answers is not in the document]
Comment (ohadbitt) [reply azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py:23]: nice!
does this return a generator?

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def create_fast_suffix_matcher_from_existing(rules: List[MatchRule], existing: FastSuffixMatcher) -> FastSuffixMatcher:
+    if existing is None or len(existing.rules) == 0:
+        return FastSuffixMatcher(rules)
+
+    if not rules:
+        return existing
+
+    lst = copy.deepcopy(rules)
```
[the remark this answers is not in the document]
Comment (ohadbitt) [reply azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py]: values is 2d list - had to change it to:
`([*copy.deepcopy(rules), *(v for item in existing.rules.values() for v in item)])`

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+        raise KustoClientInvalidConnectionStringException(
+            f"Can't communicate with '{hostname}' as this hostname is currently not trusted; please see " f"https://aka.ms/kustotrustedendpoints"
+        )
+
+    def set_override_policy(self, matcher):
+        self._override_matcher = matcher
+
+
+_filename = os.path.join(os.path.split(os.path.abspath(__file__))[0], "wellKnownKustoEndpoints.json")
```
[the remark this answers is not in the document]
Comment (ohadbitt) [reply azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py:107]: Talked F2F - its part of the package, packed by setup.py

File: azure-kusto-data/azure/kusto/data/client.py [hunk 7]
```diff
@@ -133,6 +133,7 @@ def execute(self, database: str, query: str, properties: Optional[ClientRequestP
         :return: Kusto response data set.
         :rtype: azure.kusto.data.response.KustoResponseDataSet
         """
+        self.validate_endpoint()
```
Comment (AsafMah) [inline azure-kusto-data/azure/kusto/data/client.py:136]: You can remove all of these like in the aio client, and only have it in _execute

File: azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply azure-kusto-data/azure/kusto/data/kusto_trusted_endpoints.py]: (len(rule.suffix) for rule in it1) returns a generator, and then min() iterates it and returns the result

**Review by AsafMah — COMMENTED** [review]
One comment

**Review by AsafMah — APPROVED** [review]

**Pull request merged**